orchestrator/ingest/ingest.go

```diff
@@ -0,0 +1,186 @@
+
+// ingestionManagerInstance A singleton instance of the ingestion manager object
+var ingestionManagerInstance *IngestionManager
+
+// ack Ack sent by worker pool
+const ack = "ack"
+
+// IngestionManagerInstance A function to return an ingestion manager instance
+func IngestionManagerInstance() *IngestionManager {
+	ingestionManagerOnce.Do(func() {
+		configManager := config.ConfigurationManagerInstance("config/config_files")
+		configObj := configManager.IngestionManagerConfig("ingestion_manager.yaml")
+		params := params.OrchestratorParamsInstance()
+
+		connection, err := tcp.NewConnection(zmq4.REQ, "")
+		errors.HandleError(err, fmt.Sprintf("%s %s\n", logPrefix, "Unable to establish tcp connection"), true)
+
+		activeRoutines := 1
+
+		manager := IngestionManager{
+			workerPoolIP:           configObj.WorkerPoolIP,
+			workerPoolPort:         configObj.WorkerPoolPort,
+			connectionHandler:      connection,
+			startIdx:               params.StartIdx,
+			frameCount:             params.FrameCount,
+			jobSize:                configObj.JobSize,
+			workersScaningInterval: time.Duration(configObj.WorkersScanningInterval) * time.Second,
+			jobSendTimeout:         time.Duration(configObj.JobSendTimeout) * time.Second,
+			jobsListMutex:          &sync.Mutex{},
+			workersListMutex:       &sync.Mutex{},
+			activeJobsMutex:        &sync.Mutex{},
+			activeRoutines:         activeRoutines,
+		}
+
+		manager.populateJobsPool()
+
+		ingestionManagerInstance = &manager
+	})
+
+	return ingestionManagerInstance
+}
+
+// Start Starts the ingestion manager job scheduling routine
+func (manager *IngestionManager) Start() {
+	log.Println(logPrefix, "Starting Ingestion Manager")
+
+	manager.wg.Add(manager.activeRoutines)
+
+	//Start jobs assigner
+	go manager.assignJobs()
+}
+
+// Shutdown A function to shutdown the ingestion manager
+func (manager *IngestionManager) Shutdown() {
+	log.Println(logPrefix, "Shutting down ingestion manager")
+
+	for i := 0; i < manager.activeRoutines; i++ {
+		manager.shutdown <- true
+	}
+
+	log.Println(logPrefix, "Waiting for ingestion routines to terminate")
+	manager.wg.Wait()
+
+	log.Println(logPrefix, "Health check monitor shutdown successfully")
+}
+
+func (manager *IngestionManager) assignJobs() {
+	defer manager.wg.Done()
+
+	log.Println(logPrefix, "Jobs Assigner Started")
+
+	manager.establishConnection()
+
+	for range time.Tick(manager.workersScaningInterval) {
+		select {
+		case <-manager.shutdown:
+			log.Println(logPrefix, "Ingestion Manager is shutting down")
+
+			return
+		default:
+			if len(manager.jobsList) == 0 {
+				continue
+			}
+
+			manager.jobsListMutex.Lock()
+			manager.workersListMutex.Lock()
+			manager.activeJobsMutex.Lock()
+
+			//Assign jobs to non-busy workers
+			for _, worker := range manager.workers {
+				if worker.Utilization.Busy {
+					continue
+				}
+
+				//Check todo jobs
+				for _, job := range manager.jobsList {
+					go manager.sendJob(job)
```

Comment:
why loop if you break after the first element? just fetch the head and send it as a job

Reply:
It's a map. I don't there is a short cut for getting the first element in the map.